_posts/2021-06-26-complete-guide-to-cors.md

```diff
@@ -0,0 +1,441 @@
+
+The browser does this by exchanging a set of CORS headers with the server. Based on the header values returned from the server, the browser provides access to the server response or blocks the access by throwing a CORS error. 
+
+### Important CORS Headers
+
+When we make a CORS request, the browser sends a header named `Origin` with the request to the server. To serve this cross-origin request, the server sends back a header named `Access-Control-Allow-Origin` in the response. 
+
+The browser checks the value of the `Access-Control-Allow-Origin` header in the response and renders the response only if the value of the `Access-Control-Allow-Origin` header is the same as the `Origin` header sent in the request. The server can also use wild cards like `*`  as the value of the `Access-Control-Allow-Origin` header to represent a partial match with the value of the `Origin` header received in the request.
+
+### CORS Failures
+
+CORS failures cause errors but specifics about the error are not available to the browser for security reasons. The only way to know about the error is by looking at the browser's console for details of the error which is usually in the following form:
+
+```shell
+Access to XMLHttpRequest at 'http://localhost:8000/orders' from origin 'http://localhost:9000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
+```
+
+## Type of CORS Requests
+
+The browser sends three types of CORS requests: `simple`, `preflight`, and `requests with credentials`. The browser determines the type of request to be sent to the server depending on the kind of operations we want to perform with the server resource. Let us understand these request types and observe these requests in the browsers' network log by running an example.
+
+### Simple CORS Requests (GET, POST, and HEAD)
+Simple requests are sent by the browser for server operations it considers safe because they do not change the state of any existing resource on the server. The request sent by the browser is simple if one of the below conditions applies: 
+
+- The HTTP request method is GET, POST, or HEAD
+- The HTTP request contains a CORS safe-listed header: Accept, Accept-Language, Content-Language, Content-Type.
+- When using the Content-Type header, the only values allowed are: application/x-www-form-urlencoded, multipart/form-data, or text/plain
+- No event listeners are registered on any [XMLHttpRequestUpload](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/upload) object
+- No ReadableStream object is used in the request
+
+The browser sends the simple request as a normal request similar to the Same Origin request, and the `Access-Control-Allow-Origin` header is checked by the browser when the response is returned.
+
```

Comment:
So, simple cross-origin requests are ALWAYS allowed? I didn't quite gather that from the text.